rust/connlib/connection/src/allocation.rs

```diff
@@ -0,0 +1,890 @@
+
+                self.allocation_lifetime = Some((now, lifetime));
+                update_candidate(
+                    maybe_srflx_candidate,
+                    &mut self.last_srflx_candidate,
+                    &mut self.new_candidates,
+                );
+                update_candidate(
+                    maybe_ip4_relay_candidate,
+                    &mut self.last_ip4_candidate,
+                    &mut self.new_candidates,
+                );
+                update_candidate(
+                    maybe_ip6_relay_candidate,
+                    &mut self.last_ip6_candidate,
+                    &mut self.new_candidates,
+                );
+
+                tracing::info!(
+                    srflx = ?self.last_srflx_candidate,
+                    relay_ip4 = ?self.last_ip4_candidate,
+                    relay_ip6 = ?self.last_ip6_candidate,
+                    ?lifetime,
+                    "Updated candidates of allocation"
+                );
+            }
+            REFRESH => {
+                let Some(lifetime) = message.get_attribute::<Lifetime>() else {
+                    tracing::warn!("Message does not contain lifetime");
+                    return true;
+                };
+
+                self.allocation_lifetime = Some((now, lifetime.lifetime()));
+
+                tracing::info!(
+                    srflx = ?self.last_srflx_candidate,
+                    relay_ip4 = ?self.last_ip4_candidate,
+                    relay_ip6 = ?self.last_ip6_candidate,
+                    ?lifetime,
+                    "Updated lifetime of allocation"
+                );
+            }
+            CHANNEL_BIND => {
+                let Some(channel) = original_request
+                    .get_attribute::<ChannelNumber>()
+                    .map(|c| c.value())
+                else {
+                    tracing::warn!("Request did not contain a `CHANNEL-NUMBER`");
+                    return true;
+                };
+
+                if !self.channel_bindings.set_confirmed(channel, now) {
+                    tracing::warn!(%channel, "Unknown channel");
+                };
+
+                return true;
+            }
+            _ => {}
+        }
+
+        true
+    }
+
+    /// Attempts to decapsulate and incoming packet as a channel-data message.
+    ///
+    /// Returns the original sender, the packet and _our_ remote socket that this packet was sent to.
+    pub fn decapsulate<'p>(
+        &mut self,
+        from: SocketAddr,
+        packet: &'p [u8],
+        now: Instant,
+    ) -> Option<(SocketAddr, &'p [u8], SocketAddr)> {
+        if from != self.server {
+            return None;
+        }
+
+        let (peer, payload) = self.channel_bindings.try_decode(packet, now)?;
+
+        // Our remote socket that we received this packet on!
+        let remote_socket = match peer {
+            SocketAddr::V4(_) => self.last_ip4_candidate.as_ref()?.addr(),
```

Comment:
`last_ip4_candidate` candidate is our IPv4 allocation address. I guess the "last" is somewhat confusing meaning because there is always only one and we always refresh it so it can never get stale. Technically though, in TURN, the allocated IP can expire and we will then get a new one if we make another allocation.

Reply:
This also isn't the IP we are speaking with but it is _our_ remote socket. I guess "remote" is also bad naming here. It is remote in that it isn't our local socket. I'll think of some better names.